lib/resources/sampleBuffer.js

```diff
@@ -0,0 +1,80 @@
+const os = require('node:os')
+const crypto = require('crypto')
+
+const instanceId = crypto.createHash('md5').update(os.hostname()).digest('hex').substring(0, 4)
+class SampleBuffer {
+    constructor (size = 1000) {
+        this.size = size
+        this.buffer = new Array(this.size)
+        this.head = 0
+        this.wrapped = false
+        this.lastTimestamp = 0
+        this.lastTimestampCount = 0
+    }
+
+    add (sample) {
+        if (!sample.ts) {
+            sample.ts = Date.now()
+        }
+        this.buffer[this.head++] = sample
+        if (this.head == this.size) {
+            this.head = 0
+            this.wrapped = true
+        }
+        return sample
+    }
+
+    clear () {
+        this.buffer = new Array(this.size)
+        this.head = 0
+        this.wrapped = false
+    }
+
+    toArray () {
+        if (!this.wrapped) {
+            return this.buffer.slice(0, this.head)
+        } else  {
+            const result = this.buffer.slice(this.head, this.size)
+            result.push(...this.buffer.slice(0,this.head))
+            return result
+        }
+    }
+
+    lastX (x) {
+        if (this.head > x) {
+            return this.buffer.slice(this.head - x, this.head)
+        } else {
+            if (this.wrapped) {
+                const d = x - this.head
+                const result = this.buffer.slice(this.size - d, this.size)
+                result.push(...this.buffer.slice(0, this.head))
+                return result
+            } else {
+                return this.buffer.slice(0, this.head)
+            }
+        }
+    }
+
+    avgLastX (x) {
+        const samples  = this.lastX(x)
+        const result = {}
+        samples.forEach(sample => {
+            for (const [key, value] of Object.entries(sample)) {
+                if (key !== 'ts' && key !== 'err') {
```

Comment:
Here we skip samples with errors. which is going to skew the average down. We could remove 1 from the sample count for each skipped sample, which given the current sample average period is a lot longer than unhealthy time out this should work out.